### PR TITLE
Add localizedTime() validation rule

### PR DESCRIPTION
**Pull request by ypnos-web**
Functionality described in #8323 

Test is still missing. The test would go along the lines of creating I18n::Time and I18n::Date objects of various dates and using their string output as input to the rule next to invalid strings.

However I first ask for feedback regarding how to write proper tests.

### CODE REVIEW

File: src/Validation/Validation.php [hunk 1]
```diff
@@ -457,6 +458,33 @@ public static function time($check)
...
+        }
+        static $methods = [
+            'date' => 'parseDate',
+            'time' => 'parseTime',
+            'datetime' => 'parseDateTime',
+        ];
+        if (empty($methods[$type])) {
+            throw new \InvalidArgumentException('Unsupported parser type given.');
+        }
+        return (!is_null(Time::{$methods[$type]}($check, $format)));
```
Comment (dereuromark) [inline src/Validation/Validation.php]: is_null should be `!== null`

File: src/Validation/Validation.php [hunk 2]
```diff
@@ -457,6 +458,33 @@ public static function time($check)
     }
 
     /**
+     * Date and/or time string validation.
+     * Uses `I18n::Time` to parse the date. This means parsing is locale dependent.
+     *
+     * @param string|\DateTime $check a date string or object (will always pass)
+     * @param string $type Parser type, one out of 'date', 'time', and 'datetime'
+     * @param string|int $format any format accepted by IntlDateFormatter
```
Comment (dereuromark) [inline src/Validation/Validation.php]: |null missing

File: src/Validation/Validation.php [hunk 3]
```diff
@@ -457,6 +458,33 @@ public static function time($check)
...
+     * @return bool Success
+     * @throws \InvalidArgumentException when unsupported $type given
+     * @see \Cake\I18N\Time::parseDate(), \Cake\I18N\Time::parseTime(), \Cake\I18N\Time::parseDateTime()
+     */
+    public static function localizedTime($check, $type = 'datetime', $format = null)
+    {
+        if ($check instanceof DateTime) {
```
Comment (markstory) [inline src/Validation/Validation.php]: You might want to check `DateTimeInterface` instead to cover DateTimeImmutable.

**Comment by markstory** [comment]
Code looks good. For tests I would add the following cases, for each of Date/Time/DateTime:

* Invalid value
* Valid value
* Valid but incorrectly formatted date based on the locale.
* Empty string

**Comment by lorenzo** [comment]
ping @ypnos-web

**Comment by ypnos-web** [comment]
Hello, I am currently on vacation, I plan to finish work (tests) on this PR in a week.

File: src/Validation/Validation.php (hunk 3 above)
Comment (ypnos-web) [inline src/Validation/Validation.php]: The same check is present in the original date/time methods, might be worth revisiting them as well.

**Comment by ypnos-web** [comment]
I am finished with tests and suggest to merge.

**Comment by dereuromark** [comment]
I like it!

**Comment by markstory** [comment]
Looks like [there is a test failure](https://travis-ci.org/cakephp/cakephp/jobs/116112884). Other than failing tests this is looking good.

**Comment by ypnos-web** [comment]
This is the test that `12.90am` should not be a time (but PHP does create a time object out of it). It seems I got confused when I tested the tests.

Fixed.

**Pull request merged**